exec/java-exec/src/main/java/org/apache/drill/exec/server/options/SystemOptionManager.java

```diff
@@ -70,11 +69,13 @@
  *  </p>
  */
 
-public class SystemOptionManager extends BaseOptionManager implements OptionManager, AutoCloseable {
+public class SystemOptionManager extends BaseOptionManager implements AutoCloseable {
   private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(SystemOptionManager.class);
 
-  private CaseInsensitiveMap<OptionValidator> VALIDATORS;
-  public void populateValidators() {
+  public static final CaseInsensitiveMap<OptionValidator> DEFAULT_VALIDATORS =
```

Comment:
This change reverses a change made by Jyothsna. While it seems reasonable to create a static table of validators, there is a subtle problem which Jyothsna's fix addressed.

Under normal operation, we have exactly one Drillbit per process and thus exactly one system option manager.

But, we have unit tests that create multiple Drillbits per process. This is the most convenient way to test concurrency issues in unit tests. In general, each Drillbit could define a different default value for the system/session option; perhaps to check forward/backward compatibility when the time comes to support rolling upgrades.

In this model, each Drillbit will have a slightly different set of defaults. Thus, the option validator table must be a member variable of the system option manager; it cannot be global.

Note also, in the existing tests with multiple Drillbits, a global table runs into a race condition during reading defaults from the config file if we use a single table. But, if we have a separate table per Drillbit, the race condition disappears.

Reply:
Agreed we should not be using global variables. The intended flow was the following:

- DEFAULT_VALIDATORS holds the default set of validators
- The system option manager is passed a set of validators when it is initialized
- The validators map is copied and stored in the validators variable on line 230. So each option manager has a separate copy of the validator map.

Looking at this again I agree that this is not ideal. The DEFAULT_VALIDATORS constant should be deleted and usages should be replaced with createDefaultValidators.

I also see another issue that was present before where the validators themselves are static, they should be explicitly copied, so we should probably add a clone method to the validators to do that.